cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java

```diff
@@ -0,0 +1,268 @@
+      if (!namespaceId.equals(NamespaceId.SYSTEM.getEntityName())) {
+        LocationIdentifier logLocationIdentifier = locationManager.getLocationIdentifier(eventObject
+                                                                                           .getMDCPropertyMap());
+        rollover(logLocationIdentifier, eventObject);
+        OutputStream locationOutputStream = locationManager.getLocationOutputStream(logLocationIdentifier, filePath);
+        setOutputStream(locationOutputStream);
+        writeOut(eventObject);
+      }
+    } catch (IllegalArgumentException iae) {
+      // this shouldn't happen
+      LOG.error("Unrecognized context ", iae);
+    } catch (IOException ioe) {
+      throw new LogbackException("Exception while appending event. ", ioe);
+    }
+  }
+
+  private void rollover(final LocationIdentifier logLocationIdentifier, ILoggingEvent logEvent) {
+    if (!locationManager.getActiveLocations().containsKey(logLocationIdentifier)) {
+      return;
+    }
+    final LocationOutputStream locationOutputStream = locationManager.getActiveLocations().get(logLocationIdentifier);
+
+    if (triggeringPolicy instanceof LocationTriggeringPolicy) {
+      // no need to type cast on every event
+      if (locationTriggeringPolicy == null) {
+        locationTriggeringPolicy = ((LocationTriggeringPolicy) triggeringPolicy);
+      }
+
+      locationTriggeringPolicy.setLocation(locationOutputStream.getLocation());
+      // set number of bytes written to locationOutputStream, we need to do this because HDFS does not provide
+      // correct size of the file
+      locationTriggeringPolicy.setActiveLocationSize(locationOutputStream.getNumOfBytes());
+
+      if (locationTriggeringPolicy.isTriggeringEvent(logEvent)) {
+        try {
+          if (rollingPolicy instanceof LocationRollingPolicy) {
+            // no need to type cast on every event
+            if (locationRollingPolicy == null) {
+              locationRollingPolicy = ((LocationRollingPolicy) rollingPolicy);
+            }
+
+            locationRollingPolicy.setLocation(locationOutputStream.getLocation(), new Closeable() {
+              @Override
+              public void close() throws IOException {
+                locationOutputStream.close();
+                locationManager.getActiveLocations().remove(logLocationIdentifier);
+              }
+            });
+
+            locationRollingPolicy.rollover();
+          }
+        } catch (RolloverFailure e) {
+          // we do not want to stop processing because roll over failed. so catch it and process the event
+          LOG.warn("Attempt to rollover failed for appender {}.", name);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void flush() throws IOException {
+    locationManager.flush();
+  }
+
+  @Override
+  public void stop() {
+    if (encoder != null) {
+      try {
+        LOG.info("Stopping appender {}", this.name);
+        encoder.close();
```

Comment:
would it be safer to do `encoder.init(null)` before calling close ? otherwise, encoder will close one of the output stream (latest one it inited) and we want to manage flushing/closing output streams.